src/main/scala/extracells/util/FluidUtil.java

```diff
@@ -87,20 +102,135 @@ public static MutablePair<Integer, ItemStack> fillStack(
 		return null;
 	}
 
+	/**
+	 * Try to fill a liquid container item from AE fluid storage. Make sure to SIMULATE first.
+	 * @param container Item to fill, stackSize must be 1
+	 * @param request Requested fluid amount, the real amount drained may be smaller
+	 * @param monitor The AE network
+	 * @param mode Whether to simulate or actually move liquids
+	 * @param src Action source for auditing
+	 * @return Null if failed, otherwise the filled container and the fluid stack it was filled with
+	 */
+	public static MutablePair<ItemStack, FluidStack> fillItemFromAe(ItemStack container, FluidStack request, IMEMonitor<IAEFluidStack> monitor, Actionable mode, BaseActionSource src) {
+		if (container == null || container.stackSize != 1) {
+			return null;
+		}
+		if (request == null || request.amount <= 0) {
+			return null;
+		}
+		if (monitor == null) {
+			return null;
+		}
+		int capacity = FluidUtil.getCapacity(container, request.getFluid());
+		int requestAmount = Integer.min(capacity, request.amount);
+		IAEFluidStack result = monitor.extractItems(FluidUtil.createAEFluidStack(request.getFluid(), requestAmount), mode, src);
+		if (result == null || result.getStackSize() <= 0) {
+			return null;
+		}
+		if (!result.getFluid().equals(request.getFluid())) {
+			FMLLog.severe("[ExtraCells2] ME network returned fluid `%s` when requesting `%s`",
```

Comment:
That may look paranoid but this is actually a valid concern, because the fluid may be returned from a 3rd-party storage handled by a 3rd party storage handler, so, gj I guess.